### PR TITLE
UsageStats: Add connected users and client

### PR DESCRIPTION
**Pull request by torkelo**
This adds two new metrics to usage stats

* live_users
* live_clients

Stats taken from the websocket centrifuge servers. Currently these are always the same, so same user in two tabs counts as different users. @FZambia do you know why?

### CODE REVIEW

File: pkg/services/live/live.go [hunk 1]
```diff
@@ -724,6 +720,13 @@ func (g *GrafanaLive) ClientCount(orgID int64, channel string) (int, error) {
 	return len(p.Presence), nil
 }
 
+func (g *GrafanaLive) UsageStats() UsageStats {
+	clients := g.node.Hub().NumClients()
+	users := g.node.Hub().NumClients()
```
Comment (DanCech) [inline pkg/services/live/live.go:725]:
```suggestion
	users := g.node.Hub().NumUsers()
```

**Comment by DanCech** [comment]
This is a really cool idea, but I'm worried that the data isn't going to be useful because it'll be sampled only once per day and it's much more real-time than say active users (or even a daily active users count) so the numbers will be highly dependent on the alignment between when you take the sample and the active hours of the users.  One possibility that comes to mind would be to accumulate these stats over time and report back something like min/max/avg to have some more useful data.

File: pkg/services/live/live.go (hunk 1 above)
Comment (torkelo) [reply pkg/services/live/live.go:725]: ah that explains why these where same! LOL, :facepalm:

**Comment by torkelo** [comment]
> This is a really cool idea, but I'm worried that the data isn't going to be useful because it'll be sampled only once per day and it's much more real-time than say active users (or even a daily active users count) so the numbers will be highly dependent on the alignment between when you take the sample and the active hours of the users. One possibility that comes to mind would be to accumulate these stats over time and report back something like min/max/avg to have some more useful data.

Ah, that is very true. Think checking every 10min and having min/max/avg is a good idea, and reset those every 24h

**Comment by DanCech** [comment]
> Ah, that is very true. Think checking every 10min and having min/max/avg is a good idea, and reset those every 24h

Exactly, the main question in my mind is who should be responsible for that accumulation over time.  If it's UsageStats then we can relatively easily accumulate min/max/avg and just reset them when the report is sent.  If it's GrafanaLive then we would need to keep the individual readings so that it can calculate the last-24hrs stats whenever UsageStats needs them.

**Comment by torkelo** [comment]
@DanCech pushed a quick first draft of a sampling code (from usage stats service), as it already has timers for stats sampling/collection feel best to have it this way. 

It was a bit of a rush job (not tested it), need to drive to pickup kid now. If you want to refactor/refine it please push to this branch.

**Comment by DanCech** [comment]
Awesome, will check it out!

**Comment by FZambia** [comment]
> Stats taken from the websocket centrifuge servers. Currently these are always the same, so same user in two tabs counts as different users. @FZambia do you know why?

Hi, is this still actual? I just tried to reproduce this with code like this inside Live:

```go
	go func() {
		for {
			time.Sleep(5 * time.Second)
			println(node.Hub().NumClients(), node.Hub().NumUsers())
		}
	}()
```

Opened 2 tabs: num clients - 2 and num users - 1.

**Comment by torkelo** [comment]
@FZambia sorry that comment was before Dan noticed I had  stupid typo in the code (always used NumClients)

File: pkg/infra/usagestats/usage_stats.go [hunk 2]
```diff
@@ -282,6 +297,8 @@ func (uss *UsageStatsService) sendUsageStats(ctx context.Context) error {
 	data := bytes.NewBuffer(out)
 	sendUsageStats(data)
 
+	uss.resetLiveStats()
```
Comment (marefr) [inline pkg/infra/usagestats/usage_stats.go:300]: If `!uss.Cfg.ReportingEnabled` is true this code will not execute - is that expected? 

Maybe `sampleLiveStats()` should only run if `uss.Cfg.ReportingEnabled` is true?

**Review by marefr — APPROVED** [review]
LGTM, but left a comment

File: pkg/infra/usagestats/usage_stats_test.go [hunk 3]
```diff
@@ -608,5 +612,13 @@ func createService(t *testing.T, cfg setting.Cfg) *UsageStatsService {
...
+		grafanaLive:        newTestLive(t),
 	}
 }
+
+func newTestLive(t *testing.T) *live.GrafanaLive {
+	cfg := &setting.Cfg{AppURL: "http://localhost:3000/"}
+	gLive, err := live.ProvideService(nil, cfg, routing.NewRouteRegister(), nil, nil, nil, nil, sqlstore.InitTestDB(t))
+	require.NoError(t, err)
+	return gLive
+}
```
[the remark this answers is not in the document]
Comment (FZambia) [reply pkg/infra/usagestats/usage_stats_test.go:624]: Yeah, ugly - will try to address this soon.

File: pkg/infra/usagestats/usage_stats.go [hunk 4]
```diff
@@ -77,6 +77,21 @@ func (uss *UsageStatsService) GetUsageReport(ctx context.Context) (UsageReport,
 	metrics["stats.dashboards_viewers_can_admin.count"] = statsQuery.Result.DashboardsViewersCanAdmin
 	metrics["stats.folders_viewers_can_edit.count"] = statsQuery.Result.FoldersViewersCanEdit
 	metrics["stats.folders_viewers_can_admin.count"] = statsQuery.Result.FoldersViewersCanAdmin
+	metrics["stats.folders_viewers_can_admin.count"] = statsQuery.Result.FoldersViewersCanAdmin
```
Comment (FZambia) [inline pkg/infra/usagestats/usage_stats.go:80]: This line looks as a duplicate of the line above.

**Review by FZambia — APPROVED** [review]

File: pkg/infra/usagestats/usage_stats.go (hunk 4 above)
Comment (torkelo) [reply pkg/infra/usagestats/usage_stats.go]: ah, thanks for spotting this, will remove

File: pkg/infra/usagestats/usage_stats.go (hunk 2 above)
Comment (torkelo) [reply pkg/infra/usagestats/usage_stats.go]: @marefr did a minor refactor to move this to the ticker instead

File: pkg/infra/usagestats/usage_stats.go [hunk 5]
```diff
@@ -264,9 +278,6 @@ func (uss *UsageStatsService) RegisterMetricsFunc(fn MetricsFunc) {
 }
 
 func (uss *UsageStatsService) sendUsageStats(ctx context.Context) error {
-	if !uss.Cfg.ReportingEnabled {
-		return nil
-	}
```
Comment (marefr) [inline pkg/infra/usagestats/usage_stats.go:281]:
```suggestion
```

**Review by marefr — APPROVED** [review]
LGTM, but left a suggestion to resolve lint issues

**Review by DanCech — APPROVED** [review]
This looks great!

**Comment by torkelo** [comment]
Sadly this was messy to backport, (due to wire migration, and live v8.2 work), so moving this to v8.2

**Comment by DanCech** [comment]
@torkelo probably cleaner to have stats updates in 8.2 anyway

**Pull request merged**